### PR TITLE
Add EsScanNode

### PR DESCRIPTION
**Pull request by Salieri1969**
Add a scan node that scans es

### CODE REVIEW

File: be/src/exec/es_scan_node.h [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+    // scan range's iterator, used in get_next()
+    int _scan_range_idx;
+
+    // store every scan range's netaddress/handle/offset
+    std::vector<TNetworkAddress> _addresses;
+    std::vector<std::string> _scan_handles;
+    std::vector<int> _offsets;
+
+    Status open_es(TNetworkAddress& address, TExtOpenResult& result, TExtOpenParams& params);
```
Comment (imay) [inline be/src/exec/es_scan_node.h]: put these functions before member variables

File: be/src/exec/es_scan_node.cpp [hunk 2]
```diff
@@ -0,0 +1,660 @@
...
+    if (tnode.es_scan_node.__isset.properties) {
+        _properties = tnode.es_scan_node.properties;
+    }
+}
+
+EsScanNode::~EsScanNode() {
+}
+
+Status EsScanNode::prepare(RuntimeState* state) {
+    LOG(INFO) << "EsScanNode::Prepare";
```
Comment (imay) [inline be/src/exec/es_scan_node.cpp]: This log should not be a INFO level

File: be/src/exec/es_scan_node.cpp [hunk 3]
```diff
@@ -0,0 +1,660 @@
...
+
+    RETURN_IF_ERROR(ScanNode::prepare(state));
+    _tuple_desc = state->desc_tbl().get_tuple_descriptor(_tuple_id);
+    if (nullptr == _tuple_desc) {
```
Comment (imay) [inline be/src/exec/es_scan_node.cpp]:
```suggestion
     if (_tuple_desc == nullptr) {
```

compiler can help us

File: be/src/exec/es_scan_node.cpp [hunk 4]
```diff
@@ -0,0 +1,660 @@
...
+        LOG(WARNING) << ss.str();
+        return Status(ss.str());
+    }
+    _env = state->exec_env();
+
+    return Status::OK;
+}
+
+Status EsScanNode::open(RuntimeState* state) {
+    LOG(INFO) << "EsScanNode::Open";
```
Comment (imay) [inline be/src/exec/es_scan_node.cpp]: LOG level

File: be/src/exec/es_scan_node.cpp [hunk 5]
```diff
@@ -0,0 +1,660 @@
...
+    SCOPED_TIMER(_runtime_profile->total_time_counter());
+    RETURN_IF_ERROR(ExecNode::open(state));
+
+    // TExtOpenParams.row_schema
+    vector<TExtColumnDesc> cols;
+    for (const SlotDescriptor* slot : _tuple_desc->slots()) {
+        TExtColumnDesc col;
+        col.__set_name(slot->col_name());
+        col.__set_type(slot->type().to_thrift());
+        cols.push_back(std::move(col));
```
Comment (imay) [inline be/src/exec/es_scan_node.cpp]:
```suggestion
         cols.emplace_back(std::move(col));
```

File: be/src/exec/es_scan_node.cpp [hunk 6]
```diff
@@ -0,0 +1,660 @@
...
+        }
+        return true;
+    } else {
+        VLOG(1) << "get disjuncts fail: node type is " << conjunct->node_type()
+                << ", should be BINARY_PRED or COMPOUND_PRED";
+        return false;
+    }
+}
+
+TExtLiteral EsScanNode::get_literal(ExprContext* context, Expr* expr) {
```
Comment (imay) [inline be/src/exec/es_scan_node.cpp]:
```suggestion
TExtLiteral EsScanNode::to_exe_literal(ExprContext* context, Expr* expr) {
```

File: be/src/exec/es_scan_node.cpp [hunk 7]
```diff
@@ -0,0 +1,660 @@
...
+
+Status EsScanNode::materialize_row(MemPool* tuple_pool, Tuple* tuple,
+                                   const vector<TExtColumnData>& cols, int row_idx,
+                                   vector<int>& cols_next_val_idx) {
+  tuple->init(_tuple_desc->byte_size());
+
+  for (int i = 0; i < _tuple_desc->slots().size(); ++i) {
+    const SlotDescriptor* slot_desc = _tuple_desc->slots()[i];
+    void* slot = tuple->get_slot(slot_desc->tuple_offset());
+    const TExtColumnData& col = cols[i];
```
Comment (imay) [inline be/src/exec/es_scan_node.cpp]: If slot_desc->is_materialized() is false, you should skip this slot

File: be/src/exec/es_scan_node.cpp [hunk 8]
```diff
@@ -0,0 +1,660 @@
...
+                val_size, "string slot");
+            return tuple_pool->mem_tracker()->MemLimitExceeded(NULL, details, val_size);
+          }
+          memcpy(buffer, val.data(), val_size);
+          reinterpret_cast<StringValue*>(slot)->ptr = buffer;
+          reinterpret_cast<StringValue*>(slot)->len = val_size;
+          break;
+        }
+      case TYPE_TINYINT:
+        if (val_idx >= col.byte_vals.size()) {
```
Comment (imay) [inline be/src/exec/es_scan_node.cpp]: why isn't size() col's function?

File: be/src/exec/es_scan_node.cpp [hunk 9]
```diff
@@ -0,0 +1,660 @@
...
+
+  for (int i = 0; i < _tuple_desc->slots().size(); ++i) {
+    const SlotDescriptor* slot_desc = _tuple_desc->slots()[i];
+    void* slot = tuple->get_slot(slot_desc->tuple_offset());
+    const TExtColumnData& col = cols[i];
+
+    if (col.is_null[row_idx]) {
+      tuple->set_null(slot_desc->null_indicator_offset());
+      continue;
+    }
```
Comment (imay) [inline be/src/exec/es_scan_node.cpp]: else {
tuple->set_not_null()
}

**Review by imay — CHANGES_REQUESTED** [review]

File: be/src/exec/es_scan_node.cpp (hunk 8 above)
Comment (Salieri1969) [reply be/src/exec/es_scan_node.cpp]: struct TExtColumnData {
  1: required list<bool> is_null;
  2: optional list<bool> bool_vals;
  3: optional binary byte_vals;
  4: optional list<i16> short_vals;
  5: optional list<i32> int_vals;
  6: optional list<i64> long_vals;
  7: optional list<double> double_vals;
  8: optional list<string> string_vals;
  9: optional list<binary> binary_vals;
}

File: be/src/exec/es_scan_node.cpp [hunk 10]
```diff
@@ -0,0 +1,667 @@
...
+        if (get_disjuncts(_conjunct_ctxs[i], _conjunct_ctxs[i]->root(), disjuncts)) {
+            predicates.push_back(std::move(disjuncts));
+            conjunct_idxes.push_back(i);
+        }
+    }
+
+    // open every scan range
+    int conjunct_accepted_times[_conjunct_ctxs.size()]; 
+    for (int i = 0; i < _scan_ranges.size(); ++i) {
+        TEsScanRange es_scan_range = _scan_ranges[i];
```
Comment (imay) [inline be/src/exec/es_scan_node.cpp]: reference

File: be/src/exec/es_scan_node.cpp [hunk 11]
```diff
@@ -0,0 +1,667 @@
...
+        params.__set_row_schema(row_schema);
+        params.__set_batch_size(state->batch_size());
+        params.__set_predicates(predicates);
+        TExtOpenResult result;
+
+        // check es host
+        if (es_scan_range.es_hosts.empty()) {
+            std::stringstream ss;
+            ss << "es fail to open: hosts empty";
+            LOG(ERROR) << ss.str();
```
Comment (imay) [inline be/src/exec/es_scan_node.cpp]: use WARNING instead, ERROR means serious things happened

File: be/src/exec/es_scan_node.cpp [hunk 12]
```diff
@@ -0,0 +1,667 @@
...
+                if (j == selected_idx) continue;
+                es_host_selected = es_scan_range.es_hosts[j];
+                status = open_es(es_host_selected, result, params);
+                if (status.code() == TStatusCode::ES_SHARD_NOT_FOUND) {
+                    continue;
+                } else {
+                    break;
+                }
+            }
+        }
```
Comment (imay) [inline be/src/exec/es_scan_node.cpp]: you can use 

```
for (int i = 0; i < 2; ++i) {
   for (auto& host : es_scan_range.es_host) {
        if (i == 0 && host != localhost) {
         continue;
        } else (i == 1 &&  j == selected_idx) {
         continue;
        }
    }
}
```

File: be/src/exec/es_scan_node.cpp [hunk 13]
```diff
@@ -0,0 +1,667 @@
...
+        VLOG(1) << "conjunct: " << _conjunct_ctxs[i]->root()->debug_string();
+        vector<TExtPredicate> disjuncts;
+        if (get_disjuncts(_conjunct_ctxs[i], _conjunct_ctxs[i]->root(), disjuncts)) {
+            predicates.push_back(std::move(disjuncts));
+            conjunct_idxes.push_back(i);
+        }
+    }
+
+    // open every scan range
+    int conjunct_accepted_times[_conjunct_ctxs.size()];
```
Comment (imay) [inline be/src/exec/es_scan_node.cpp]: without initialize

use std::vector<int> instead

**Review by imay — CHANGES_REQUESTED** [review]

File: be/src/exec/es_scan_node.cpp [hunk 14]
```diff
@@ -128,66 +136,59 @@ Status EsScanNode::open(RuntimeState* state) {
...
-                if (status.code() == TStatusCode::ES_SHARD_NOT_FOUND) {
+        // choose an es node, local is the first choice
+        std::string localhost = BackendOptions::get_localhost();
+        bool is_success = false;
+        for (int j = 0; j < 2; ++j) {
+            for (auto& es_host : es_scan_range.es_hosts) {
+                if ((j == 0 && es_host.hostname != localhost)
+                    || (j == 1 && es_host.hostname == localhost)) {
                     continue;
                 } else {
```
Comment (imay) [inline be/src/exec/es_scan_node.cpp]: there is no need to have a `else`

File: be/src/exec/es_scan_node.cpp [hunk 15]
```diff
@@ -0,0 +1,668 @@
...
+        DCHECK(scan_range.scan_range.__isset.es_scan_range);
+        TEsScanRange es_scan_range = scan_range.scan_range.es_scan_range;
+        _scan_ranges.push_back(es_scan_range);
+    }
+
+    _offsets.resize(scan_ranges.size(), 0);
+    return Status::OK;
+}
+
+Status EsScanNode::open_es(TNetworkAddress& address, TExtOpenResult& result, TExtOpenParams& params) {
```
Comment (imay) [inline be/src/exec/es_scan_node.cpp]: input argument use `const reference` and output argument use `pointer`

File: be/src/exec/es_scan_node.cpp [hunk 16]
```diff
@@ -0,0 +1,668 @@
...
+#else
+    TStatus status;
+    result.__set_status(status);
+    result.__set_scan_handle("0");
+    return Status(status);
+#endif
+}
+
+bool EsScanNode::get_disjuncts(ExprContext* context, Expr* conjunct,
+                               vector<TExtPredicate>& disjuncts) {
```
Comment (imay) [inline be/src/exec/es_scan_node.cpp]: output use pointer, and input use const reference

File: be/src/exec/es_scan_node.cpp [hunk 17]
```diff
@@ -0,0 +1,668 @@
...
+        VLOG(1) << "get disjuncts fail: node type is " << conjunct->node_type()
+                << ", should be BINARY_PRED or COMPOUND_PRED";
+        return false;
+    }
+}
+
+TExtLiteral EsScanNode::to_exe_literal(ExprContext* context, Expr* expr) {
+    void* value = context->get_value(expr, NULL);
+    TExtLiteral literal;
+    literal.__set_node_type(expr->node_type());
```
Comment (imay) [inline be/src/exec/es_scan_node.cpp]: ExprContext ctx
expr.get_xxx_value(&ctx, nullptr)

File: be/src/exec/es_scan_node.cpp [hunk 18]
```diff
@@ -0,0 +1,668 @@
...
+            op = conjunct->op();
+        } else if (TExprNodeType::SLOT_REF == conjunct->get_child(1)->node_type()) {
+            expr = conjunct->get_child(0);
+            slotRef = (SlotRef*)(conjunct->get_child(1));
+            op = conjunct->op();
+        } else {
+            VLOG(1) << "get disjuncts fail: no SLOT_REF child";
+            return false;
+        }
+
```
Comment (imay) [inline be/src/exec/es_scan_node.cpp]: you need to check if `expr` is a literal

**Review by imay — APPROVED** [review]

**Pull request merged**